File: lifelink/mobile/lib/domain/entities/blood_request_entity.dart
class BloodRequestEntity {
  final String id;
  final String patientName;
  final String bloodGroup;
  final int unitsRequired;
  final String hospitalName;
  final String contactNumber;
  final String? requiredDate;
  final String urgency;
  final String status;
  final String? city;
  final double? latitude;
  final double? longitude;
  final String? distance;
  final String? requesterName;
  final String createdAt;

  const BloodRequestEntity({
    required this.id,
    required this.patientName,
    required this.bloodGroup,
    required this.unitsRequired,
    required this.hospitalName,
    required this.contactNumber,
    this.requiredDate,
    required this.urgency,
    required this.status,
    this.city,
    this.latitude,
    this.longitude,
    this.distance,
    this.requesterName,
    required this.createdAt,
  });

  factory BloodRequestEntity.fromJson(Map<String, dynamic> json) => BloodRequestEntity(
        id: json['id'],
        patientName: json['patient_name'],
        bloodGroup: json['blood_group'],
        unitsRequired: json['units_required'] ?? 1,
        hospitalName: json['hospital_name'],
        contactNumber: json['contact_number'],
        requiredDate: json['required_date'],
        urgency: json['urgency'] ?? 'Normal',
        status: json['status'] ?? 'Open',
        city: json['city'],
        latitude: double.tryParse(json['latitude']?.toString() ?? ''),
        longitude: double.tryParse(json['longitude']?.toString() ?? ''),
        distance: json['distance_km'] != null ? '${json['distance_km']} km' : null,
        requesterName: json['requester']?['name'],
        createdAt: json['created_at'] ?? DateTime.now().toIso8601String(),
      );
}

File: lifelink/backend/src/database/entities/blood-request.entity.ts
import {
  Entity, PrimaryGeneratedColumn, Column, CreateDateColumn,
  UpdateDateColumn, ManyToOne, JoinColumn, Index,
} from 'typeorm';
import { User } from './user.entity';
export { Urgency, RequestStatus } from './enums';
import { BloodGroup, Urgency, RequestStatus } from './enums';

@Entity('blood_requests')
export class BloodRequest {
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Column({ length: 255 })
  patient_name: string;

  @Index()
  @Column({ type: 'enum', enum: BloodGroup })
  blood_group: BloodGroup;

  @Column({ type: 'int', default: 1 })
  units_required: number;

  @Column({ length: 255 })
  hospital_name: string;

  @Column({ length: 20 })
  contact_number: string;

  @Column({ type: 'date', nullable: true })
  required_date: Date;

  @Index()
  @Column({ type: 'enum', enum: Urgency, default: Urgency.NORMAL })
  urgency: Urgency;

  @Index()
  @Column({ type: 'enum', enum: RequestStatus, default: RequestStatus.OPEN })
  status: RequestStatus;

  @Column({ type: 'decimal', precision: 10, scale: 8, nullable: true })
  latitude: number;

  @Column({ type: 'decimal', precision: 11, scale: 8, nullable: true })
  longitude: number;

  @Index()
  @Column({ length: 100, nullable: true })
  city: string;

  @ManyToOne(() => User, (user) => user.blood_requests, { nullable: true, onDelete: 'SET NULL' })
  @JoinColumn({ name: 'requester_id' })
  requester: User;

  @Column({ nullable: true })
  requester_id: string;

  @CreateDateColumn()
  created_at: Date;

  @UpdateDateColumn()
  updated_at: Date;
}
